README.md
qz-print
========

Free browser applet for sending documents and raw commands to a printer.

##Getting Started
  * See our [Getting Started](../../wiki/getting-started) guide.
  * Visit our home page https://qz.io.
  
##Support
  * File a bug via our [issue tracker](../../issues)
  * Ask the community via our [community support page](https://qz.io/support/)
  * Ask the developers via [premium support](https://qz.io/contact/) (fees may apply)

##Changelog
  * See our [most recent releases](../../releases)

##Java Developer Resources
  * [Install dependencies](../../wiki/install-dependencies)
  * [Compile, Package](../../wiki/compiling)


CHANGELOG.txt
 Changes 1.9.5 (12/16/2015)
 - Adds single instance support (#95)
 - Adds mimetype URL launcher support qz:launch (#94)
 - Adds warning dialog when "Always Block" is selected (#90)
 - Adds option to reduce console.log(...) verbosity (#111)
 - Adds localhost bindings to be configurable (run as print server) (#91)
 - Adds ability to call deployQZ(...) several times (#107)
 - Adds autenticode signing to Windows build process (#80)
 - Adds continuous integration via Travis-CI (b8717a8)
 - Adds perl signing example (49dfe7c)
 - Adds override support for Trusted Root Certificate (6970815)
 - Adds support for PortableFirefox (93838ed)
 - Fixes permissions for Linux installer (3f48eee)
 - Fixes desktop icon on Linux, Mac (3d623ce2, e4b7718)
 - Fixes start menu shortcut on Windows 10 (#106)
 - Fixes printToFile(...), printToHost(...) bugs (#104, af42696, 4049cf7)
 - Fixes clearException() not working as expected (#103)
 - Fixes "Open File Location" on some versions of Linux (#116)
 - Fixes serial support for Legacy Java Applet (#100)
Changes 1.9.4 (9/16/2015)
 - Fix applet fallback on Safari (#34)
 - Fixes occasional blank security dialogs (#58)
 - Better searching for Firefox installation on Mac/Linux (#81, #85, #88)
 - Better Firefox certificate removal on Mac/Linux (3b7b8f3)
 - Allow better 4x6 paper size hints (#78)
 - Add warning when "permanently block" is selected (#90)
 - Compatibility fix for legacy Prototype.js JSON.stringify() (#76)
 - Fixes a hard crash on OSX 10.10 with monochrome images (#75)
 - Support for Microsoft Edge (#70)
 - Fixes Port Busy errors when serials ports are not closed manually (#73)
 - Fixes installation on Windows when only a 64-bit Java install is present (#64)
 - Add warning when Java isn't installed on Apple (c96ef90)
 - Fixes tray menu placement on Fedora Linux (#51)
 - Adds support for when no system tray is available/kiosk mode (#60)
 - Re-enables print orientation for PrintPostScript use (6d8e679)
 - Suppresses zero-length signature warnings from logs (7265640)
 - Increases raw buffer size from 512 to 8192 (ec72091)
Changes 1.9.3 (7/1/2105)
 - Improved signing examples, added Node.js, J#, Java, Classic ASP
 - Fix Windows installer issue when being run as non-admin use
Changes 1.9.2 (6/16/2015)
 - Better HTTPS certificate installation for Firefox (#46)
Changes 1.9.1 (6/15/2015)
 - Adds improved applet fallback support when using websockets (#34)
 - Prevent race condition when appending data via async: false (#41)
 - Fixes loading issue with Internet Explorer (#39)
 - Fixes installation problems on Apple (#37)
 - Fixes issue with printToHost() (#33)
 - Adds silent installer option for Linux via sudo qz-tray-1.9.x.run -- -y
 - Searches for certutil tool on Linux, attempts to install if missing (needed for Chrome)
Changes 1.9.0 (6/3/2015)
 - Fixes findPrinter() when printer name starts with "\\"
 - Added better messaging when certificate is invalid
 - Certificate Revocation List (CRL) switched to HTTPS, qz.io domain
 - Added CRL fetching status to the LogDialog
 - Removed duplicate log entries
 - Windows:  Fix broken uninstaller on 64-bit OS
 - Windows:  Better compatibility for installer JScript
 - Windows:  Added firewall exception logic (8181, 8182, ...)
 - Apple:  Better menu integration with OS X
 - Linux:  Added secure websocket support
Changes 1.8.94 (5/18/2015)
 - Added secure websocket support for Apple
 - Added secure websocket support for Firefox (Windows and Apple only)
 - Improved certificate cleanup on Apple
 - Prevent unnecessary signing of preemptive messages
 - Fix "Open folder location" on Apple
 - Detect and block legacy WebSockets versions
 - About dialog: Display all active ports and secure/insecure status
 - New ConfirmDialog, bring GatewayDialog to top
Changes 1.8.91 (5/2/2015)
 - Added QZ Tray support for HTTPS pages via Secure WebSockets
Changes 1.8.90 (3/16/2015)
 - Initial WebSockets release, works without NPAPI
Changes 1.8.11 (1/23/2015)
 - Added separate_jvm applet parameter
Changes 1.8.10 (1/5/2015)
 - Added rotation support via qz.setRotation(angle);
 - Remove hanging window on printPS() error (GitHub issue #40)
Changes 1.8.7 (7/22/2014)
- Fixes lockup introduced with Chrome 36.0 (GitHub issue #33)
Changes 1.8.5 (6/24/2014)
- Fixes qz.setCopies() function for PostScript printing
Changes 1.8.2 (4/23/2014)
- Fixes invalid hex character decoding bug (google code issue 191)
Changes 1.8.1 (4/15/2014)
- Fixed PostScript sizing bug when printing to two separate page sizes
- Removed separate Java 6 build
Changes 1.8.0 (12/05/2013)
- printToFile() now blocks common executable file extensions
- Fixed printHTML() typo in sample.html
- Fixed jzebraReady() backwards compatiblity for Ubuntu
- Added workaround for possible browser lock-up, documented here:
   https://bugs.launchpad.net/ubuntu/+source/firefox-3.0/+bug/331962
- IE11 Fix - Updated deployJava.js from Oracle.
- Updated jquery to latest version for HTML5 printing.
Changes 1.7.8.1 (11/16/13)
- Removed feature per request - 3rd party reflection class for PDFs
Changes 1.7.8: (11/13/13)
- Removed untrusted dialogs for Java 6 by using custom deployJava.js script
Changes 1.7.7: (11/10/13)
- Fixed whitespace character breaking JNLP on Ubuntu
- Better examples and explainations of new features in sample.html
- Remove all while() loop examples from sample.html.
- Fixed a bug where getPrinter() returns the last valid printer name after none has been found
Changes 1.7.6: (11/08/13)
- Finished MAC Address feature
- Added detailed warning message for outdated Java versions
Changes 1.7.5: (11/06/13)
- Updated qz.getMacAddress() to better detect active MAC address for interfaces with only an IPv6 LinkLocal addresses
Changes 1.7.4: (11/04/13)
- Misc branding changes in sample.html (logos, etc)
- New applet parameter <param name="initial_focus" value="false"/> added to fix focus
  issues with key listeners and barcode scanners.

js/qz-websocket.js
var qzConfig = {
    callbackMap: {
        findPrinter:     'qzDoneFinding',
        findPrinters:    'qzDoneFinding',
        appendFile:      'qzDoneAppending',
        appendXML:       'qzDoneAppending',
        appendPDF:       'qzDoneAppending',
        appendImage:     'qzDoneAppending',
        print:           'qzDonePrinting',
        printPS:         'qzDonePrinting',
        printHTML:       'qzDonePrinting',
        printToHost:     'qzDonePrinting',
        printToFile:     'qzDonePrinting',
        findPorts:       'qzDoneFindingPorts',
        openPort:        'qzDoneOpeningPort',
        closePort:       'qzDoneClosingPort',
        findNetworkInfo: 'qzDoneFindingNetwork'
    },
    protocols: ["wss://", "ws://"],   // Protocols to use, will try secure WS before insecure
    uri: "localhost",                // Base URL to server
    ports: [8181, 8282, 8383, 8484], // Ports to try, insecure WS uses port (ports[x] + 1)
    keepAlive: (60 * 1000),           // Interval in millis to send pings to server
    debug: true,

    port: function() { return qzConfig.ports[qzConfig.portIndex] + qzConfig.protocolIndex; },
    protocol: function() { return qzConfig.protocols[qzConfig.protocolIndex]; },
    url: function() { return qzConfig.protocol() + qzConfig.uri + ":" + qzConfig.port(); },
    increment: function() {
        if (++qzConfig.portIndex < qzConfig.ports.length) {
            return true;
        }
        return false;
    },
    outOfBounds: function() { return qzConfig.portIndex >= qzConfig.ports.length },
    init: function(){
        qzConfig.preemptive = {isActive: '', getVersion: '', getPrinter: '', getLogPostScriptFeatures: ''};
        qzConfig.protocolIndex = window.location.protocol == "https:" ? 0 : 1;         // Used to track which value in 'protocol' array is being used
        qzConfig.portIndex = 0;             // Used to track which value in 'ports' array is being used
        return qzConfig;
    }
};

var logger = {
    info: function(v) { console.log(v); },
    log: function(v) { if (qzConfig.debug) { console.log(v); } },
    warn: function(v) { console.warn(v); },
    error: function(v) { console.error(v); }
}

function deployQZ(host, debug) {
    if (host) {
        qzConfig.uri = host;
    }

    if (debug === false) {
        qzConfig.debug = false;
    }

    logger.log(WebSocket);
    qzConfig.init();

    // Old standard of WebSocket used const CLOSED as 2, new standards use const CLOSED as 3, we need the newer standard for jetty
    if ("WebSocket" in window && WebSocket.CLOSED != null && WebSocket.CLOSED > 2) {
        logger.info('Starting deploy of qz');
        connectWebsocket();
    } else {
        alert("WebSocket not supported");
        window["deployQZ"] = null;
    }
}

function connectWebsocket() {
    logger.log('Attempting connection on port ' + qzConfig.port());

    try {
        var websocket = new WebSocket(qzConfig.url());
    }
    catch(e) {
        logger.error(e);
    }

    if (websocket != null) {
        websocket.valid = false;

        websocket.onopen = function(evt) {
            logger.log('Open:');
            logger.log(evt);

            websocket.valid = true;
            connectionSuccess(websocket);

            // Create the QZ object
            createQZ(websocket);

            // Send keep-alive to the websocket so connection does not timeout
            // keep-alive over reconnecting so server is always able to send to client
            websocket.keepAlive = window.setInterval(function() {
                websocket.send("ping");
            }, qzConfig.keepAlive);
        };

        websocket.onclose = function(event) {
            try {
                if (websocket.valid || qzConfig.outOfBounds()) {
                    qzSocketClose(event);
                }
                // Safari compatibility fix to raise error event
                if (!websocket.valid && navigator.userAgent.indexOf('Safari') != -1 && navigator.userAgent.indexOf('Chrome') == -1) {
                    websocket.onerror();
                }
                websocket.cleanup();
            } catch (ignore) {}
        };

        websocket.onerror = function(event) {
            if (websocket.valid || qzConfig.outOfBounds()) {
                qzSocketError(event);
            }

            // Move on to the next port
            if (!websocket.valid) {
                websocket.cleanup();
                if (qzConfig.increment()) {
                    connectWebsocket();
                } else {
                    qzNoConnection();
                }
            }
        };
		
        websocket.cleanup = function() {
            // Explicitly clear setInterval
            if (websocket.keepAlive) {
                window.clearInterval(websocket.keepAlive);        		
            }
            websocket = null;
        };

    } else {
        logger.warn('Websocket connection failed');
        qzNoConnection();
    }
}

// Prototype-safe JSON.stringify
function stringify(o) {
    if (Array.prototype.toJSON) {
        logger.warn("Overriding Array.prototype.toJSON");
        var result = null;
        var tmp = Array.prototype.toJSON;
        delete Array.prototype.toJSON;
        result = JSON.stringify(o);
        Array.prototype.toJSON = tmp;
        return result;
    }
    return JSON.stringify(o);
}

function connectionSuccess(websocket) {
    logger.info('Websocket connection successful');

    websocket.sendObj = function(objMsg) {
        var msg = stringify(objMsg);

        logger.log("Sending " + msg);
        var ws = this;

        // Determine if the message requires signing
        if (objMsg.method === 'listMessages' || Object.keys(qzConfig.preemptive).indexOf(objMsg.method) != -1) {
            ws.send(msg);
        } else {
            signRequest(msg,
                        function(signature) {
                            ws.send(signature + msg);
                        }
            );
        }
    };

    websocket.onmessage = function(evt) {
        var message = JSON.parse(evt.data);

        if (message.error != undefined) {
            logger.error(message.error);
            return;
        }

        // After we ask for the list, the value will come back as a message.
        // That means we have to deal with the listMessages separately from everything else.
        if (message.method == 'listMessages') {
            // Take the list of messages and add them to the qz object
            mapMethods(websocket, message.result);

        } else {
            // Got a return value from a call
            logger.log('Message:');
            logger.log(message);

            if (typeof message.result == 'string') {
                //unescape special characters
                message.result = message.result.replace(/%5C/g, "\\").replace(/%22/g, "\"");

                //ensure boolean strings are read as booleans
                if (message.result == "true" || message.result == "false") {
                    message.result = (message.result == "true");
                }

                if (message.result.substring(0, 1) == '[') {
                    message.result = JSON.parse(message.result);
                }

                //ensure null is read as null
                if (message.result == "null") {
                    message.result = null;
                }
            }

            if (message.callback != 'setupMethods' && message.result != undefined && message.result.constructor !== Array) {
                message.result = [message.result];
            }

            // Special case for getException
            if (message.method == 'getException') {
                if (message.result != null) {
                    var result = message.result;
                    message.result = {
                        getLocalizedMessage: function() {
                            return result;
                        }
                    };
                }
            }

            if (message.callback == 'setupMethods') {
                logger.log("Resetting function call");
                logger.log(message.result);
                qz[message.method] = function() {
                    return message.result;
                }
            }

            if (message.callback != null) {
                try {
                    logger.log("Callbacking: " + message.callback);
                    if (window["qz"][message.callback] != undefined) {
                        window["qz"][message.callback].apply(this, message.init ? [message.method] : message.result);
                    } else {
                        window[message.callback].apply(this, message.result);
                    }
                }
                catch(err) {
                    logger.error(err);
                }
            }
        }

        logger.log("Finished processing message");
    };
}

function createQZ(websocket) {
    // Get list of methods from websocket
    getCertificate(function(cert) {
        websocket.sendObj({method: 'listMessages', params: [cert]});
        window["qz"] = {};
    });
}

function mapMethods(websocket, methods) {
    logger.log('Adding ' + methods.length + ' methods to qz object');
    for(var x = 0; x < methods.length; x++) {
        var name = methods[x].name;
        var returnType = methods[x].returns;
        var numParams = methods[x].parameters;

        // Determine how many parameters there are and create method with that many
        (function(_name, _numParams, _returnType) {
            //create function to map function name to parameter counted function
            window["qz"][_name] = function() {
                var func = undefined;
                if (typeof arguments[arguments.length - 1] == 'function') {
                    func = window["qz"][_name + '_' + (arguments.length - 1)];
                } else {
                    func = window["qz"][_name + '_' + arguments.length];
                }

                func.apply(this, arguments);
            };

            //create parameter counted function to include overloaded java methods in javascript object
            window["qz"][_name + '_' + _numParams] = function() {
                var args = [];
                for(var i = 0; i < _numParams; i++) {
                    args.push(arguments[i]);
                }

                var cb = arguments[arguments.length - 1];
                var cbName = _name + '_callback';

                if ($.isFunction(cb)) {
                    var method = cb.name;

                    // Special case for IE, which does not have function.name property ..
                    if (method == undefined) {
                        method = cb.toString().match(/^function\s*([^\s(]+)/)[1];
                    }

                    if (method == 'setupMethods') {
                        cbName = method;
                    }

                    window["qz"][cbName] = cb;
                } else {
                    logger.log("Using mapped callback " + qzConfig.callbackMap[_name] + "() for " + _name + "()");
                    cbName = qzConfig.callbackMap[_name];
                }

                logger.log("Calling " + _name + "(" + args + ") --> CB: " + cbName + "()");
                websocket.sendObj({method: _name, params: args, callback: cbName, init: (cbName == 'setupMethods')});
            }
        })(name, numParams, returnType);
    }

    // Re-setup all functions with static returns
    for(var key in qzConfig.preemptive) {
        window["qz"][key](setupMethods);
    }

    logger.log("Sent methods off to get rehabilitated");
}

function setupMethods(methodName) {
    if ($.param(qzConfig.preemptive).length > 0) {
        logger.log("Reset " + methodName);
        delete qzConfig.preemptive[methodName];

        logger.log("Methods left to return: " + $.param(qzConfig.preemptive).length);

        // Fire ready method when everything on the QZ object has been added
        if ($.param(qzConfig.preemptive).length == 0) {
            qzReady();
        }
    }
}
